routes/dao_coin_exchange_with_fees.go

```diff
@@ -0,0 +1,1602 @@
+	}
+
+	var txn *lib.MsgDeSoTxn
+	if IsDesoPkid(coinPublicKey) {
+		txn, _, _, _, _, err = fes.CreateSendDesoTxn(
+			int64(amountBaseUnits.Uint64()),
+			transactorPubkeyBytes,
+			receiverPubkeyBytes,
+			nil,
+			minFeeRateNanosPerKb,
+			nil)
+		if err != nil {
+			return nil, fmt.Errorf("HandleMarketOrder: Problem creating transaction: %v", err)
+		}
+	} else {
+		txn, _, _, _, err = fes.blockchain.CreateDAOCoinTransferTxn(
+			transactorPubkeyBytes,
+			&lib.DAOCoinTransferMetadata{
+				ProfilePublicKey:       coinPkBytes,
+				ReceiverPublicKey:      receiverPubkeyBytes,
+				DAOCoinToTransferNanos: *amountBaseUnits,
+			},
+			// Standard transaction fields
+			minFeeRateNanosPerKb, fes.backendServer.GetMempool(), nil)
+		if err != nil {
+			return nil, fmt.Errorf("HandleMarketOrder: Problem creating transaction: %v", err)
+		}
+	}
+
+	return txn, nil
+}
+
+func (fes *APIServer) HandleMarketOrder(
+	isMarketOrder bool,
+	req *DAOCoinLimitOrderWithFeeRequest,
+	isBuyOrder bool,
+	feeMapByPubkey map[string]uint64,
+) (
+	*DAOCoinLimitOrderWithFeeResponse,
+	error,
+) {
+	quoteCurrencyUsdValue := float64(0.0)
+	quoteCurrencyUsdValueStr, err := fes.GetQuoteCurrencyPriceInUsd(
+		req.QuoteCurrencyPublicKeyBase58Check)
+	if err != nil {
+		// If we can't get the price of the quote currency in usd, then we can't
+		// convert the usd amount to a quote currency amount. In this case, keep
+		// going but don't use the quote currency usd value for anything.
+		quoteCurrencyUsdValue = 0.0
+	} else {
+		quoteCurrencyUsdValue, err = strconv.ParseFloat(quoteCurrencyUsdValueStr, 64)
+		if err != nil {
+			// Again, get the usd value on a best-effort basis
+			quoteCurrencyUsdValue = 0.0
+		}
+	}
+	convertToUsd := func(quoteAmountStr string) string {
+		quoteAmount, err := strconv.ParseFloat(quoteAmountStr, 64)
+		if err != nil {
+			return ""
+		}
+		return fmt.Sprintf("%.9f", quoteAmount*quoteCurrencyUsdValue)
+	}
+
+	quantityStr := req.Quantity
+	if req.QuantityCurrencyType == CurrencyTypeUsd {
+		// In this case we want to convert the usd amount to an amount of quote
+		// currency in base units. To do this we need to get the price of the
+		// quote currency in usd and then convert the usd amount to quote currency
+		// amount.
+		if quoteCurrencyUsdValue == 0.0 {
+			return nil, fmt.Errorf("HandleMarketOrder: Quote currency price in " +
+				"usd not available. Please use quote or base currency for the amount.")
+		}
+		// For the rest it's just the following formula:
+		// = usd amount / quoteCurrencyUsdValue * base units
+
+		// In this case we parse the quantity as a simple float since its value
+		// should not be extreme
+		quantityUsd, err := strconv.ParseFloat(req.Quantity, 64)
+		if err != nil {
+			return nil, fmt.Errorf("HandleMarketOrder: Problem converting quantity "+
+				"to float %v", err)
+		}
+		quantityStr = fmt.Sprintf("%0.9f", quantityUsd/quoteCurrencyUsdValue)
+	}
+
+	priceStrQuote := ""
+	if !isMarketOrder {
```

Comment:
Kinda weird that this function is called `HandleMarketOrder` and it takes in a value called `isMarketOrder`. Should this function be called `HandleOrder` or something?